source-postgres/snapshot.go

```diff
@@ -0,0 +1,239 @@
+	if err := s.Transaction.Rollback(ctx); err != nil {
+		s.Conn.Close(ctx)
+		return err
+	}
+	return s.Conn.Close(ctx)
+}
+
+func (s *TableSnapshot) TransactionLSN() pglogrepl.LSN {
+	return s.DB.TransactionLSN()
+}
+
+func (s *TableSnapshot) buildScanQuery(start bool) string {
+	// We cache the query used in start=false cases (ScanFrom) to avoid
+	// redundantly building it over and over for no reason.
+	if !start && s.scanFromQuery != "" {
+		return s.scanFromQuery
+	}
+
+	// Construct strings like `(foo, bar, baz)` and `($1, $2, $3)` for use in the query
+	var pkey, args string
+	for idx, colName := range s.ScanKey {
+		if idx > 0 {
+			pkey += ", "
+			args += ", "
+		}
+		pkey += colName
+		args += fmt.Sprintf("$%d", idx+1)
+	}
+
+	// Construct the query itself
+	query := new(strings.Builder)
+	fmt.Fprintf(query, "SELECT * FROM %s.%s", s.SchemaName, s.TableName)
+	if !start {
+		fmt.Fprintf(query, " WHERE (%s) > (%s)", pkey, args)
+	}
+	fmt.Fprintf(query, " ORDER BY (%s)", pkey)
+	fmt.Fprintf(query, " LIMIT %d;", SnapshotChunkSize)
+
+	// Cache where applicable and return
+	if !start {
+		s.scanFromQuery = query.String()
+		return s.scanFromQuery
+	}
+	return query.String()
+}
+
+func (s *TableSnapshot) ScanStart(ctx context.Context, handler ChangeEventHandler) (int, []byte, error) {
+	logrus.WithFields(logrus.Fields{
+		"namespace":  s.SchemaName,
+		"table":      s.TableName,
+		"primaryKey": s.ScanKey,
+		"txLSN":      s.TransactionLSN(),
+	}).Info("starting table scan")
+
+	query := s.buildScanQuery(true)
+	logrus.WithField("query", query).Debug("executing query")
+	rows, err := s.DB.Conn.Query(ctx, query)
+	if err != nil {
+		return 0, nil, errors.Wrapf(err, "unable to execute query %q", query)
+	}
+	defer rows.Close()
+	return s.dispatchResults(ctx, rows, handler)
+}
+
+func (s *TableSnapshot) ScanFrom(ctx context.Context, prevKey []byte, handler ChangeEventHandler) (int, []byte, error) {
+	logrus.WithFields(logrus.Fields{
+		"namespace":  s.SchemaName,
+		"table":      s.TableName,
+		"primaryKey": s.ScanKey,
+		"resumeKey":  base64.StdEncoding.EncodeToString(prevKey),
+	}).Debug("scanning next chunk")
+
+	args, err := unpackTuple(prevKey)
+	if err != nil {
+		return 0, nil, errors.Wrap(err, "error unpacking encoded tuple")
+	}
+	if len(args) != len(s.ScanKey) {
+		return 0, nil, errors.Errorf("expected %d primary-key values but got %d", len(s.ScanKey), len(args))
+	}
+	query := s.buildScanQuery(false)
+	logrus.WithField("query", query).WithField("args", args).Debug("executing query")
+	rows, err := s.DB.Conn.Query(ctx, query, args...)
+	if err != nil {
+		return 0, prevKey, errors.Wrapf(err, "unable to execute query %q", query)
+	}
+	defer rows.Close()
+	return s.dispatchResults(ctx, rows, handler)
+}
+
+func (s *TableSnapshot) dispatchResults(ctx context.Context, rows pgx.Rows, handler ChangeEventHandler) (int, []byte, error) {
+	cols := rows.FieldDescriptions()
+	var lastKey []byte
+	var rowsProcessed int
+	for rows.Next() {
```

Comment:
Optional / for-later optimization questions to avoid extra allocations:
* Can |fields| be re-used in between rows ?
* Could you use rows.Values() for the first row only, and thereafter use rows.Scan() with that same value slice ?
* |keyFields| could be re-used.
* The tuple-order sanity check *could* be hoisted out, so we only encode the final key of the row.
   But I agree the sanity check is too valuable to do that unless it's *really* needed.

This ⬆️ said, the performance of this connector may not be a bottle-neck at all relative to flow runtime capture performance so probably not worth much concern or effort at this stage.

Reply:
👍